opm/output/eclipse/RestartIO.cpp

```diff
@@ -716,17 +865,68 @@ namespace {
         writeFluidInPlace(value, es, write_double_arg, rstFile);
         writeTracerVectors(schedule.getUnits(), es.tracer(), value,
                            write_double_arg, rstFile);
-        writeUDQ(report_step, sim_step, schedule, udq_state, inteHD, rstFile);
 
-        writeExtraVectors(value, writeDouble);
+        if (!is_lgr_grid) {
+            writeSolutionExtra(value, schedule, udq_state, report_step, sim_step,
+                ecl_compatible_rst, inteHD, rstFile,
+                writeDorF, writeInt, writeDouble);
+        }
 
-        if (! ecl_compatible_rst) {
-            writeExtendedSolutionVectors(value, writeDorF, writeInt);
+    }
+
+    //  Writes the solution for Global grids
+    void writeSolution(const RestartValue&           value,
+                       const EclipseState&           es,
+                       const Schedule&               schedule,
+                       const UDQState&               udq_state,
+                       int                           report_step,
+                       int                           sim_step,
+                       const bool                    ecl_compatible_rst,
+                       const bool                    write_double_arg,
+                       const std::vector<int>&       inteHD,
+                       EclIO::OutputStream::Restart& rstFile)
+    {
+        writeSolutionCore(value, es, schedule, udq_state, report_step, sim_step,
+                          ecl_compatible_rst, write_double_arg, inteHD, rstFile);
+
+        const auto &grid = es.getInputGrid();
+        if (grid.is_lgr())
+        {
+            rstFile.write("LGRNAMES",grid.get_all_lgr_labels());
         }
+        rstFile.message("ENDSOL");
```

Comment:
It looks a little weird, in my opinion, to have `ENDSOL` in a different scope to `STARTSOL`.  Please pull `STARTSOL` out of `writeSolutionCore()` and put it here&ndash;and in the LGR counterpart&ndash;